mmdet/models/detectors/deformable_detr.py

```diff
@@ -1,12 +1,598 @@
+        # N, L, 4
+        proposals = proposals.sigmoid() * scale
+        # N, L, 4, 128
+        pos = proposals[:, :, :, None] / dim_t
+        # N, L, 4, 64, 2
+        pos = torch.stack((pos[:, :, :, 0::2].sin(), pos[:, :, :, 1::2].cos()),
+                          dim=4).flatten(2)
+        return pos
+
+
+class DeformableDetrTransformerEncoder(DetrTransformerEncoder):
+
+    def _init_layers(self) -> None:
+        self.layers = ModuleList()
+        for i in range(self.num_layers):
+            self.layers.append(
+                DeformableDetrTransformerEncoderLayer(**self.layer_cfg[i]))
+
+
+class DeformableDetrTransformerDecoder(DetrTransformerDecoder):
+
+    def _init_layers(self) -> None:
+        self.layers = ModuleList()
+        for i in range(self.num_layers):
+            self.layers.append(
+                DeformableDetrTransformerDecoderLayer(**self.layer_cfg[i]))
+        if self.post_norm_cfg is not None:
+            raise ValueError('There is not post_norm in '
+                             'DeformableDetrTransformerDecoder')
+
+    def forward(self,
+                query,
+                *args,
+                reference_points=None,
+                valid_ratios=None,
+                reg_branches=None,
+                **kwargs):
+        """Forward function for `TransformerDecoder`.
+
+        Args:
+            query (Tensor): Input query with shape
+                `(num_query, bs, embed_dims)`.
+            reference_points (Tensor): The reference
+                points of offset. has shape
+                (bs, num_query, 4) when as_two_stage,
+                otherwise has shape ((bs, num_query, 2).
+            valid_ratios (Tensor): The radios of valid
+                points on the feature map, has shape
+                (bs, num_levels, 2)
+            reg_branch: (obj:`nn.ModuleList`): Used for
+                refining the regression results. Only would
+                be passed when with_box_refine is True,
+                otherwise would be passed a `None`.
+
+        Returns:
+            Tensor: Results with shape [1, num_query, bs, embed_dims] when
+                return_intermediate is `False`, otherwise it has shape
+                [num_layers, num_query, bs, embed_dims].
+        """
+        output = query
+        intermediate = []
+        intermediate_reference_points = []
+        for lid, layer in enumerate(self.layers):
+            if reference_points.shape[-1] == 4:
+                reference_points_input = \
+                    reference_points[:, :, None] * \
+                    torch.cat([valid_ratios, valid_ratios], -1)[:, None]
+            else:
+                assert reference_points.shape[-1] == 2
+                reference_points_input = \
+                    reference_points[:, :, None] * \
+                    valid_ratios[:, None]
+            output = layer(
+                output,
+                *args,
```

Comment:
I prefer to expand the args and kwargs here for readability

Reply:
@jshilong  OK, we will try to expand all parameters.